pkg/bid/parser.go

```diff
@@ -0,0 +1,245 @@
+package bid
+
+import (
+	"fmt"
+
+	v2 "github.com/conductorone/baton-sdk/pb/c1/connector/v2"
+)
+
+func ParseBid(bidStr string) (BID, error) {
+	rs := &BidScanner{str: bidStr}
+
+	tType, val, err := rs.NextToken()
+	if err != nil {
+		return nil, err
+	}
+	if tType != literal || val != BidPrefix {
+		return nil, fmt.Errorf("invalid baton id prefix: %s", val)
+	}
+
+	tType, val, err = rs.NextToken()
+	if err != nil {
+		return nil, err
+	}
+	if tType != colon || val != ":" {
+		return nil, fmt.Errorf("invalid baton id prefix: %s", val)
+	}
+
+	tType, val, err = rs.NextToken()
+	if err != nil {
+		return nil, err
+	}
+	if tType != literal {
+		return nil, fmt.Errorf("invalid baton id prefix: %s", val)
+	}
+	tType, _, err = rs.NextToken()
+	if err != nil {
+		return nil, err
+	}
+	if tType != colon {
+		return nil, fmt.Errorf("invalid baton id prefix: %s", val)
+	}
+
+	var rv BID
+	switch val {
+	case ResourceBidPrefix:
+		rv, err = parseResourcePart(rs)
+	case EntitlementBidPrefix:
+		rv, err = parseEntitlementPart(rs)
+	case GrantBidPrefix:
+		rv, err = parseGrantPart(rs)
+	default:
+		return nil, fmt.Errorf("invalid baton id type: %s", val)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	tType, _, err = rs.NextToken()
+	if err != nil {
+		return nil, err
+	}
+	if tType != eof {
+		return nil, fmt.Errorf("invalid baton id: %s", bidStr)
+	}
+	return rv, nil
+}
+
+func parseResourcePart(rs *BidScanner) (*v2.Resource, error) {
+	// "bid:r:group/sales/user/george"
+	//        ^
+	//        Parsing in this function starts here
+	resourceId := &v2.ResourceId{}
+	parentResourceId := &v2.ResourceId{}
+
+	tokens := []string{}
+	// Each loop grabs resource type, slash, resource id
+	for {
+		// Should be the resource type, or the parent resource type, or eof/colon (we're done parsing)
+		tType, token, err := rs.PeekToken()
+		if err != nil {
+			return nil, err
+		}
+		if tType == eof || tType == colon {
+			// Colon is in the case that this resource is part of an entitlement
+			break
+		}
+		if tType != literal {
+			return nil, fmt.Errorf("invalid baton id resource part: %s %v %v", token, rs.str, rs.index)
```

Comment:
Yeah, custom error that knows how to print the nice message + index is maybe the play instead 🤔